internal/delete/agent/execute.go

```diff
@@ -14,23 +14,97 @@
 package deleteagent
 
 import (
+	"fmt"
+
+	"github.com/eclipse-iofog/iofogctl/internal"
 	"github.com/eclipse-iofog/iofogctl/internal/config"
+	"github.com/eclipse-iofog/iofogctl/internal/execute"
+	"github.com/eclipse-iofog/iofogctl/pkg/iofog/install"
 	"github.com/eclipse-iofog/iofogctl/pkg/util"
 )
 
-func Execute(namespace, name string) error {
+type executor struct {
+	name        string
+	namespace   string
+	useDetached bool
+	soft        bool
+}
+
+func NewExecutor(namespace, name string, useDetached, soft bool) (execute.Executor, error) {
+	return executor{name: name, namespace: namespace, useDetached: useDetached, soft: soft}, nil
+}
+
+func (exe executor) GetName() string {
+	return exe.name
+}
+
+func (exe executor) Execute() error {
 	util.SpinStart("Deleting Agent")
 
-	// Get executor
-	exe, err := NewExecutor(namespace, name)
-	if err != nil {
-		return err
+	// Delete agent software first, so it can properly deprovision itself before being removed
+	// Get Agent from config
+	var agent config.Agent
+	var err error
+	if exe.useDetached {
+		agent, err = config.GetDetachedAgent(exe.name)
+	} else {
+		agent, err = config.GetAgent(exe.namespace, exe.name)
+	}
+	if err == nil {
+		if !exe.soft {
+			if util.IsLocalHost(agent.Host) {
+				if err = exe.deleteLocalContainer(); err != nil {
+					util.PrintInfo(fmt.Sprintf("Could not remove iofog-agent container %s. Error: %s\n", agent.Host, err.Error()))
+				}
+			} else {
+				if err = exe.deleteRemoteAgent(agent); err != nil {
+					util.PrintInfo(fmt.Sprintf("Could not remove iofog-agent from the remote host %s. Error: %s\n", agent.Host, err.Error()))
+				}
+			}
+		}
+		if exe.useDetached {
+			return config.DeleteDetachedAgent(exe.name)
+		}
+		if err = config.DeleteAgent(exe.namespace, exe.name); err != nil {
+			util.PrintInfo(fmt.Sprintf("Could not remove iofog-agent from iofogctl config. Error: %s\n", err.Error()))
+		} else {
+			defer config.Flush()
+		}
+	} else {
+		util.PrintInfo(fmt.Sprintf("Could not find iofog-agent in iofogctl config. Error: %s\n", err.Error()))
```

Comment:
Should it not be a fatal error?

Reply:
Also the rest of the code refers to them as 'Agent' i think, not 'iofog-agent'

Reply:
> Should it not be a fatal error?

I don't believe so, because if you have a corrupted iofogctl config file, you could still need to delete the agent from the Controller database.